crates/red_knot_python_semantic/src/semantic_index/visibility_constraints.rs

```diff
@@ -553,37 +556,99 @@ impl VisibilityConstraints {
         }
     }
 
+    fn analyze_single_pattern_predicate_kind<'db>(
+        db: &'db dyn Db,
+        predicate_kind: &PatternPredicateKind<'db>,
+        subject: Expression<'db>,
+    ) -> Truthiness {
+        match predicate_kind {
+            PatternPredicateKind::Value(value) => {
+                let subject_ty = infer_expression_type(db, subject);
+                let value_ty = infer_expression_type(db, *value);
+
+                if subject_ty.is_single_valued(db) {
+                    Truthiness::from(subject_ty.is_equivalent_to(db, value_ty))
+                } else {
+                    Truthiness::Ambiguous
+                }
+            }
+            PatternPredicateKind::Singleton(singleton) => {
+                let subject_ty = infer_expression_type(db, subject);
+
+                if subject_ty.is_single_valued(db) {
+                    Truthiness::from(match singleton {
+                        ruff_python_ast::Singleton::None => subject_ty.is_none(db),
+                        ruff_python_ast::Singleton::True => {
+                            subject_ty.is_equivalent_to(db, Type::BooleanLiteral(true))
+                        }
+                        ruff_python_ast::Singleton::False => {
+                            subject_ty.is_equivalent_to(db, Type::BooleanLiteral(false))
+                        }
+                    })
+                } else {
+                    Truthiness::Ambiguous
+                }
+            }
```

Comment:
is this branch tested at all? It seems like if I apply this diff to your branch then all tests still pass, so perhaps not?

```suggestion
            PatternPredicateKind::Singleton(singleton) => Truthiness::Ambiguous,
```

I think we can also do much better here! Consider something like this:

```py
def f(x: int):
    match x:
        case None:
            y = 42
        case _:
            y = 56
    reveal_type(y)
```

On your branch, this currently reveals `Literal[42, 56]`. But in reality `y` will always have value `56`, because `int` is disjoint from `None`, so the first branch will never be taken. We can infer `Truthiness::AlwaysFalse` for `PatternPredicateKind::Singleton::(Singleton::None)` if `subject_ty.is_disjoint_from(Type::none(db))` evaluates to `true`. And similar for the other singletons.